### PR TITLE
Move envs to 'env' dir, add ray runner, update mcts strategies

### PR DESCRIPTION
**Pull request by andyk**
@nickjalbert, **I think this is ready to commit**.

~**THIS IS WORK IN PROGRESS** -- and **not ready for commit yet**. (ray_algo.py is broken and mcts.py is broken as I'm in the middle of adding a DNN to do n-step TD).~

This PR is rebased on top of your master branch which means it is using your new performance improvements.

**All tests currently pass**

This PR...

* moves environments into an env folder and it makes Base2048/Nick2048 both gym Envs. I.e., Base2048 now inherits from gym.Env (though it is not a fully compliant gym env, see next bullet re that). This required a few small tweaks to the API, but lets us easily use existing algos/FWs like Ray or spinningup. The most significant change is that step() now returns a 4th argument which is always an empty dictionary. This required updating many tests and also updating the andy_2048 env to behave similarly so that the test_do_stats tests would pass.

* Adds Nick2048Gym as a fully Gym compliant env, which is a thin wrapper around Nick2048 that just translates board tuples to and from numpy arrays. **NOTE:** currently none of Base2048, Nick2048, are fully OpenAI gym compliant, though they use gym style `action_space`s and `observation_space`s.

* adds a ray_algo.py strategy (and the dependency on ray[rllib] in requirements.txt) that currently runs DQN but also demonstrates how to run other algos that come out of the box with Ray

* makes some updates to mcts and mcts_dynamic and leaves them both in runnable state, and runs black on them

### CODE REVIEW

**Comment by andyk** [comment]
oh crap, it looks like the standard for gym envs is to have observations be np arrays, so this version where we return tuples doesn't work.

**Comment by andyk** [comment]
I've made updates so that I think this is ready for a review, discussion, and possibly merging (pending any fixes or other changes we decide on).

File: envs/nick_gym_adapter.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class Nick2048Gym(Nick2048):
+    def reset(self):
+        new_board = super().reset()
+        return np.asarray(new_board)
+
+    def get_state(self):
+        board, score, done = super().get_state()
+        return np.asarray(board), score, done
+
+    # returns an bos
```
Comment (nickjalbert) [inline envs/nick_gym_adapter.py]: Is there a typo in this comment?

Comment (andyk) [reply envs/nick_gym_adapter.py]: fixed.

**Pull request merged**